rust/relay/src/main.rs

```diff
@@ -1,4 +1,54 @@
+mod server;
+
+use anyhow::Result;
+use server::Server;
+use tokio::net::UdpSocket;
+use tracing::level_filters::LevelFilter;
+use tracing::Level;
+use tracing_subscriber::EnvFilter;
+
+const MAX_UDP_SIZE: usize = 65536;
+
 #[tokio::main]
-async fn main() {
-    println!("Hello, world!")
+async fn main() -> Result<()> {
+    tracing_subscriber::fmt()
+        .with_env_filter(
+            EnvFilter::builder()
+                .with_default_directive(LevelFilter::INFO.into())
+                .from_env_lossy(),
+        )
+        .init();
+
+    let socket = UdpSocket::bind("0.0.0.0:3478").await?;
+
+    tracing::info!("Listening on: {addr}", addr = socket.local_addr()?);
+
+    let mut server = Server::default();
+
+    let mut buf = [0u8; MAX_UDP_SIZE];
+
+    loop {
+        // TODO: Listen for websocket commands here and update the server state accordingly.
```

Comment:
We should have 2 tasks one for the TURN/STUN messages another for websocket commands and join futures here.

Reply:
I was going to keep it in the same task but `select` between them, i.e. the main loop would:

- Listen for incoming UDP packets
- Listen for incoming websocket events

Whatever comes first triggers a state change on the server and potentially send out bytes to someone.

Reply:
We'd have to benchmark it later but at least for the moment, I am unsure if multiple tasks will help us.

1. Communication between tasks needs to use bounded-channels otherwise we don't have backpressure.
2. Handling the individual packets shouldn't be very CPU-intensive.

The only way on how I can think of splitting this is if we de-multiplex the packets by source address, i.e. have one state per client. The task reading from the UDP socket would send the read frame into a different channel per IP. The other end of the channel would run in a separate task, process the packet and send a message back to write it out via the socket.

We will have to benchmark the actual processing to see if this is worth it. If we are careful in not copying around data too much during the processing, I wouldn't be surprised if we can scale this quite far using the current architecture which is delightfully simple.

Reply:
I meant just have a separate task reading from the websocket and from the UDP socket but selecting between them is also fine, I wasn't talking about having a separate task per UDP connection or channel (we can explore if that's convenient later).

Reply:
I need `&mut` access to the `Server` struct which either requires a `Arc<Mutex>` between the two tasks or we use channels. Unless you see a big benefit I am thinking of going with the current design for simplicity :)

Reply:
Go ahead with this design 👍